File: topologies/pipeline.clj
(ns pipeline
  (:use     [streamparse.specs])
  (:gen-class))

(defn pipeline [options]
   [
    ;; spout configuration
    {
    "image-spout" (python-spout-spec
          options
          "spouts.new_images.NewImageSpout"
          ["page_url", "image_url", "products", "method"]
          )
    }
    {
    "image-bolt" (python-bolt-spec
          options
          {"image-spout" :shuffle}
          "bolts.image.NewImageBolt"
          {"person_args" ["person"]
           "image_obj" ["image_obj", "image_id"]}
          :p 5
          )

    "person-bolt" (python-bolt-spec
          options
          {["image-bolt" "person_args"] :shuffle}
          "bolts.person.PersonBolt"
          {"item_args" ["item" "person_id"]
           "person_obj" ["person_obj", "person_id", "image_id"]}
	      :p 10
          )

    "item-bolt" (python-bolt-spec
          options
          {["person-bolt" "item_args"] :shuffle}
          "bolts.item.ItemBolt"
          ["item", "person_id"]
          :p 10
          )

    "merge-items-bolt" (python-bolt-spec
          options
          {["person-bolt" "person_obj"] ["person_id"]
            "item-bolt" ["person_id"]}
          "bolts.person.MergeItems"
          ["person", "image_id"]
          :p 2
          )

    "merge-people-bolt" (python-bolt-spec
          options
          {["image-bolt" "image_obj"] ["image_id"]
            "merge-items-bolt" ["image_id"]}
          "bolts.image.MergePeople"
          []
          :p 2
          )
    }
  ]
)

File: topologies/pipeline.py
"""
Pipeline topology
"""

from streamparse import Grouping, Topology
from bolts.image import NewImageBolt, MergePeople
from bolts.person import PersonBolt, MergeItems
from bolts.item import ItemBolt
from spouts.new_images import NewImageSpout


# top topology class
class pipeline(Topology):

    # Spouts
    image_spout = NewImageSpout.spec()

    # Bolts
    image_bolt = NewImageBolt.spec(inputs=[image_spout], par=1)

    person_bolt = PersonBolt.spec(inputs=[image_bolt], par=15)

    item_bolt = ItemBolt.spec(inputs=[person_bolt], par=15)

    merge_items_bolt = MergeItems.spec(inputs={person_bolt['person_obj']: Grouping.fields('person_id'), item_bolt: Grouping.fields('person_id')}, par=2)

    merge_people_bolt = MergePeople.spec(inputs={image_bolt['image_obj']: Grouping.fields('image_id'), merge_items_bolt: Grouping.fields('image_id')}, par=2)
